app/src/main/java/com/example/wellfed/mealplan/MealPlanDB.java

```diff
@@ -1,337 +1,360 @@
+         *                 false otherwise.
+         */
+        void onGetMealPlanResult(MealPlan mealPlan, boolean success);
+    }
+
+    /**
+     * This interface is used to handle the result of
+     * updating the chosen MealPlan object in the db.
+     */
+    public interface OnUpdateMealPlanListener {
+        /**
+         * Called when updateMealPlan returns a result.
+         *
+         * @param mealPlan the MealPlan object updated in the db,
+         *                 null if the update operation failed.
+         * @param success  true if the update operation was successful,
+         *                 false otherwise.
+         */
+        void onUpdateMealPlanResult(MealPlan mealPlan, boolean success);
+    }
+
+    /**
+     * Constructor for class MealPlanDB, initializes declared fields.
+     * @param connection the DBConnection object used to access the db.
+     */
+    public MealPlanDB(DBConnection connection) {
+        // Creates new instances of RecipeDB and storageIngredientDB
+        // based on current user connection.
+        recipeDB = new RecipeDB(connection);
+        ingredientDB = new IngredientDB(connection);
+
+        // Gets the instance of the Firebase FireStore DB based
+        // on current user connection.
+        this.db = connection.getDB();
+
+        // Gets the current user's MealPlan collection from db,
+        // create one if the collection DNE.
+        this.mealPlanCollection = connection.getCollection("MealPlan");
+    }
+
+    /**
+     * This method is used to add a MealPlan object to the db.
+     * @param mealPlan the MealPlan object to be added to the db.
+     * @param listener the OnAddMealPlanListener object to handle the result.
+     */
+    public void addMealPlan(MealPlan mealPlan, OnAddMealPlanListener listener) throws InterruptedException {
+        // Stores each ingredient as a HashMap with fields:
+        // ingredientRef, amount & unit.
+        ArrayList<HashMap<String, Object>> mealPlanIngredients = new ArrayList<>();
+
+        for (Ingredient i: mealPlan.getIngredients()) {
+            // Gets ingredient from db.
+            ingredientDB.getIngredient(i, (foundIngredient, success1) -> {
+                // Initializes a mapping for the ingredient.
+                HashMap<String, Object> ingredientMap = new HashMap<>();
+
+                // If the ingredient already exists in our Ingredient collection.
+                if (foundIngredient != null) {
+                    DocumentReference ingredientRef = ingredientDB.getDocumentReference(foundIngredient);
+                    i.setId(foundIngredient.getId());
+
+                    ingredientMap.put("ingredientRef", ingredientRef);
+                    ingredientMap.put("amount", i.getAmount());
+                    ingredientMap.put("unit", i.getUnit());
+                } else {
+                    // If the ingredient does not exist in our Ingredient collection,
```

Comment:
this will create a race condition, not giving the desired results since our mealPlanIngredients might not have all the ingredients when we exit this loop.